### ISSUE TITLE
Handling NaN and inf Values for 1-D Matrix Profile Algorithms

### ISSUE THREAD

**Issue by mexxexx** [TDAmeritrade/stumpy#115]
When implementing the SDTS algorithm from the Matrix Profile IV Paper, the matrix profile of a time series containing inf values has to be computed. Such a case is not yet handled in stumpy.

In line with the paper I would propose to ignore all subsequences, where at least one value is nan or inf. The tsmp R project does this, see [here on line 98] (https://github.com/cran/tsmp/blob/master/R/stomp.R). Such a change would require (at least, as for as I have seen) three changes: 
1. Changing the calculate_mean_std function to use the np.strided rolling window approach. This should be done anyways, because then the numpy std and mean functions can be used that should be more numerically stable and optimized.
2. All algorithms that calculate the MP need to keep track of subsequences that have to be skipped and need to replace nan and inf values in the time series (see the implementation in tsmp)
3. Adding a test that contains such values. We could i.e. use the time series [1, 0, 0, inf, 1, 0, 0] with m=3 that should result in the profile [0, inf, inf, inf, 0]

**Comment by mexxexx** [comment]
I implemented this functionality and some tests for STAMP and STOMP and am working on STUMP right now. 
@seanlaw Are you interested in this functionality? Then I could issue a pull request when I'm a bit more advanced

**Comment by seanlaw** [comment]
@mexxexx Yes, we would be interested in taking a look. As you pointed out, the problem is that STUMPY doesn't work when NaN/inf exists within the time series. My hesitation is largely attributed to the fact that I haven't thought about this much but I'm open to it.

> Changing the calculate_mean_std function to use the np.strided rolling window approach. This should be done anyways, because then the numpy std and mean functions can be used that should be more numerically stable and optimized.

Would you mind elaborating a little more on this? We are using stride tricks elsewhere within the code (say, `core.rolling_window`) and I'd like to confirm this statement regarding "numerically stable and optimized". Since the standard deviation/mean are only computed once before a matrix profile is computed, I'm not sure the juice is really worth the squeeze here. Perhaps, the requirement or use case is different for SDTS? Or are you saying that the NumPy functions will handle NaN/infs?

> All algorithms that calculate the MP need to keep track of subsequences that have to be skipped and need to replace nan and inf values in the time series (see the implementation in tsmp)

My concern with performing this book keeping is code branching which adds to code complexity and reduced readability. Obviously, if our data contains a ton of NaNs/infs then we take a performance hit but I don't expect this to be generally true with our user's data. If this were to be added, I'd want it to be added in all of the functions where MP is an output. As long as the current set of unit tests pass (without change) then we should be good. Note that we really don't use `STOMP` anywhere as it is only left in the library for reference.

> Adding a test that contains such values. We could i.e. use the time series [1, 0, 0, inf, 1, 0, 0] with m=3 that should result in the profile [0, inf, inf, inf, 0]

Thank you for this test case! I think we'll want a few more tests as this can be tricky

**Comment by mexxexx** [comment]
@zpzim Thank you for this suggestion! Indeed that made the implementation pretty clean. Because of the use of fastmath which makes life hard when checking for nans, I set the mean for any 'illegal' subsequence to np.inf. I implemented it now for stamp, stomp, stump and stumped, and all tests are passing. You can have a look at it in my fork. I did not implement it yet for mstump and gpu stump algorithms. For the second one there no tests, so I decided not to mess around with it.

Also, the test cases are pretty simple right now. I will add one more test for nan values at the beginning of the series, but probably the utils.naivemass should be changed to handle these cases too. I will look into it. 

> Would you mind elaborating a little more on this? We are using stride tricks elsewhere within the code (say, core.rolling_window) and I'd like to confirm this statement regarding "numerically stable and optimized". Since the standard deviation/mean are only computed once before a matrix profile is computed, I'm not sure the juice is really worth the squeeze here. Perhaps, the requirement or use case is different for SDTS? Or are you saying that the NumPy functions will handle NaN/infs?

The easiest way to compute the rolling mean and std is calling
'''
np.mean(rolling_window(T, m), axis=1)
'''
Like this we cane use the builtin functions that are optimized and tested. I only changed how the method compute_mean_std works internally, so everything else stays exactly the same.

**Comment by mexxexx** [comment]
To give a short update:
I was able to implement the SDTS training algorithm and to recreate the trafic loop example from the paper. I plan on writing a short tutorial for this.
There are still some things missing: The prediction part and some tests, ans also the computation of Fscores is really slow, but overall I am happy that it is working. I chose a modular approach, i.e. the matrix profile algorithm can be chosen by the user.
When I resolved these issues I was thinking about creating a pull request fixing issue #115 for the one dimensional algorithms (excluding gpu stump). Do you agree @seanlaw, or do you want to wait until also the multidimensional ones support nan/inf values? I will do this in the future, but it hasn't got high priority right now for me.

**Comment by seanlaw** [comment]
@mexxexx Thanks for the update! I am okay with a PR that exclusively solves the Nan/Inf problem for 1-D algorithms. I want the PR to be for a single purpose and not combined with other issues (like SDTS - which should be its [own issue](https://github.com/TDAmeritrade/stumpy/issues/118)). What is important is the test cases. Once the test cases are all in place then it will help me understand more clearly what we are trying to solve for.

I've [started an issue](https://github.com/TDAmeritrade/stumpy/issues/117) for handling NaN/inf in `mstump`/`mstumped`/`gpu_stump`. 

At the risk of sounding OCD, we strongly prefer PRs that have a single commit that addresses a single issue so, depending on how you've structured things, we may ask you to submit a new PR accordingly in order to keep the code clean. If you'd like to submit a throw-away PR for now just to get some feedback then I'd be happy to take a look. I hope that makes sense. Let me know if you have any questions.

**Comment by seanlaw** [comment]
You probably already do this but, for any PRs, please make sure to follow Git flow:

1. Fork the latest version of stumpy
2. Create a branch that fixes one issue
3. Make changes
4. Commit the changes
5. Push changes back to your fork
6. Submit a PR
1.

**Comment by seanlaw** [comment]
@mexxexx I'm playing around with some data that contains `np.nan` and the matrix profile being returned from stumpy may be `np.inf`. Here's an example:

```
import numpy as np
import stumpy

x = np.array([np.nan, 0.72983872, 0.80553998, 0.79576039, 0.8111625, 0.19467722, 0.81146815, 0.34590066, 0.25827506, 0.69051666])

mp = stumpy.stump(x, 10)
```

This produces:

```
array([[inf, -1, -1, -1],
       [2.572003803037445, 4, -1, 4],
       [2.4451437564458582, 5, -1, 5],
       [inf, -1, -1, -1],
       [2.5720038030374455, 1, 1, -1],
       [2.4451437564458574, 2, 2, -1]], dtype=object)
```

Should that be the expected behavior (to get `np.inf`? Or is it better to set the `np.inf` to `np.nan`? The problem that I have right now is that I am trying to subtract two matrix profiles from each other (not the indices) and the `np.inf` are causing warnings:

```
mp[:, 0] - mp[:, 0]

RuntimeWarning: invalid value encountered in subtract

array([nan, 0.0, 0.0, nan, 0.0, 0.0], dtype=object)
```

It is seems to be caused by trying to subtract one `np.inf` from another `np.inf`. The only way that I could find to suppress this warning is by first converting all of the `np.inf` to `np.nan` but this is a pain 😢

**Comment by mexxexx** [comment]
I'm a bit confused by your example. You have `x.size=10` and `m=10`, but the matrix profile has length `6`? And where does the second `inf` come from?

Whould you rather always return `nan` instead of `inf` or only sometimes? And does it really make a difference? It should be only one line of code to transform everything. Alternatively, you could do something like this
`np.subtract(x, y, out=np.full(y.size, np.nan), where=~(np.isinf(x) | np.isinf(y)))`
which calculates `x-y` and everywhere where either `x` or `y` are `inf` it returns `nan`, but addmittedly its not a nice solution.

I think your error is raised by subtracting two `infs` one from another. May

**Comment by seanlaw** [comment]
> I'm a bit confused by your example. You have x.size=10 and m=10, but the matrix profile has length 6? And where does the second inf come from?

@mexxexx So sorry for the confusion, it should've said `mp = stumpy.stump(x, 5)` where `m=5`, not `m=10` (I've fixed it now). I was playing around with creating a simple example and copied over the wrong line.

> Whould you rather always return nan instead of inf or only sometimes? 

Honestly, I don't know. Previously, before we supported `np.nan`/`np.inf` (i.e., when there were no missing values), I had set the matrix profile value to `np.inf` in special circumstances so that the matrix profile could still be plotted. At that time, I had not considered setting the values to `np.nan`. I haven't thought through this thoroughly so I would like to have the discussion to see if it makes more sense to have everything returned as `np.nan` or if there are benefits to differentiating between `np.nan` and `np.inf` in the matrix profile being returned. What are your thoughts?

I'm currently working on a [Tutorial for Shapelet Discovery](https://github.com/TDAmeritrade/stumpy/blob/master/docs/Tutorial_Shapelet_Discovery.ipynb) which requires subtracting one matrix profile from another (see cell 6 and cell 7). Since I inserted `np.nan` between each time series and then computed the matrix profile, the matrix profile contains `np.inf`. To correct this, I've had to manually convert the `np.inf` to `np.nan` to correct for this. This extra conversion step was alarming to me and made it seem like the stumpy output was wrong/bad but maybe I'm just overreacting.
